### PR TITLE
gh-74044: inspect.signature for wrappers around decorated bound methods

### PR DESCRIPTION
**Pull request by anton-ryzhov**
If we wrap function with bound method, which is also a wrapper around function, `inspect.signature` will not do `skip_bound_arg`.
It will use `inspect.unwrap` and pass by bound method from outer function to inner one.

Reproducer:

```python
import functools, inspect


def decorator(func):
    @functools.wraps(func)
    def inner(*args):
        return func(*args)
    return inner


class Foo(object):
    @decorator
    def bar(self, testarg):
        pass


f = Foo()
baz = decorator(f.bar)
assert inspect.signature(baz) == inspect.signature(f.bar)
```

* Issue: gh-74044

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow these steps to rectify the issue:

1. Sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username" requested by the form is the "Login name" field in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
2. **Wait at least one US business day** and then check the "Contributor form received entry under "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
3. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

**Comment by anton-ryzhov** [comment]
@the-knights-who-say-ni Got "Contributor Form Received"

File: Lib/inspect.py [hunk 1]
```diff
@@ -2171,7 +2171,8 @@ def _signature_from_callable(obj, *,
 
     # Was this function wrapped by a decorator?
     if follow_wrapper_chains:
-        obj = unwrap(obj, stop=(lambda f: hasattr(f, "__signature__")))
```
Comment (1st1) [inline Lib/inspect.py:2445]: Need to add a comment here explaining the "or" in unwrap lambda.

**Comment by 1st1** [comment]
@anton-ryzhov Good catch. The code looks OK, just add a comment explaining why your change works in all cases.

**Comment by anton-ryzhov** [comment]
@1st1 In case you haven't received notification about last push, I'm notifying about this by the comment.

**Comment by vstinner** [comment]
I removed the " needs backport to 3.6" label, the 3.6 branch no longer accept bugfixes (only security fixes are accepted): https://devguide.python.org/#status-of-python-branches

File: Lib/inspect.py (hunk 1 above)
Comment (anton-ryzhov) [reply Lib/inspect.py:2445]: If you've missed this change — I've added this in commit below. Should this branch be resolved to continue review of this PR?

**Comment by csabella** [comment]
@1st1, I believe your requested changes have been made and this is ready for another review.  Thanks!

**Comment by csabella** [comment]
Closing and reopening to trigger tests.

**Review by MaxwellDupre — APPROVED** [review]
Tested as per given test case, result pass.
Module test passed test_inspect.
Noted make failed to build nis module.
Looks ok to me.

File: Lib/inspect.py [hunk 2]
```diff
@@ -2442,7 +2442,10 @@ def _signature_from_callable(obj, *,
...
+        # Unwrap until it has a explicit signature
+        # or it is a MethodType - we need to process it individually
```
Comment (iritkatriel) [inline Lib/inspect.py]:
```suggestion
        # Unwrap until we find an explicit signature or a MethodType (which will be
        # handled explicitly below).
```

**Review by iritkatriel — APPROVED** [review]

**Comment by iritkatriel** [comment]
@anton-ryzhov This needs a [news entry](https://blurb-it.herokuapp.com/add_blurb).

**Comment by anton-ryzhov** [comment]
Done

File: Misc/NEWS.d/next/Library/2022-11-09-20-48-42.gh-issue-74044.zBj26K.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+:func:`inspect.signature` gets correct arguments for decorated methods
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Library/2022-11-09-20-48-42.gh-issue-74044.zBj26K.rst]:
```suggestion
Fixed bug where :func:`inspect.signature` reported incorrect arguments for decorated methods.
```

**Pull request merged**